### PR TITLE
MAINT: Implement 'everything follows X' for BasicStatistics

### PR DESCRIPTION
**Pull request by david-cortes-intel**
## Description

Implements 'everything follows X' logic for `BasicStatistics` and `IncrementalBasicStatistics`.

Depends on https://github.com/uxlfoundation/scikit-learn-intelex/pull/3117 and currently includes the commits from it.

This is purposefully not testing `.partial_fit()` due to the logic for mixed devices not being established yet at this point, and due to issues with the computed results on GPU.

---

<details>
<summary>Checklist:</summary>

**Completeness and readability**

- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/doc/sources/contribute.rst) for details)_.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

</details>

### CODE REVIEW

File: onedal/basic_statistics/incremental_basic_statistics.py [hunk 1]
```diff
@@ -111,6 +111,10 @@ def partial_fit(self, X, sample_weight=None, queue=None):
             Training data batch, where `n_samples` is the number of samples
             in the batch, and `n_features` is the number of features.
 
+        dtype : np.dtype
+            DType of 'X', as as a NumPy dtype or as otherwise a class that
```
Comment (avolkov-intel) [inline onedal/basic_statistics/incremental_basic_statistics.py]:
```suggestion
            DType of 'X', as a NumPy dtype or as otherwise a class that
```

File: onedal/basic_statistics/incremental_basic_statistics.py [hunk 2]
```diff
@@ -131,7 +135,7 @@ def partial_fit(self, X, sample_weight=None, queue=None):
         X_table, sample_weight_table = to_table(X, sample_weight, queue=queue)
 
         if not hasattr(self, "_onedal_params"):
-            self._onedal_params = self._get_onedal_params(False, dtype=X.dtype)
```
Comment (avolkov-intel) [inline onedal/basic_statistics/incremental_basic_statistics.py:134]: In which case we can't use previous approach?

File: sklearnex/basic_statistics/tests/test_basic_statistics.py [hunk 3]
```diff
@@ -401,3 +400,26 @@ def test_results_have_underscores(underscore_first):
...
+)
+@pytest.mark.parametrize(
+    "X_xp",
+    [np, array_api_strict],
+)
+@pytest.mark.parametrize(
+    "w_xp",
+    [np, array_api_strict],
+)
+def test_bs_weights_in_different_namespace(X_xp, w_xp, with_array_api):
```
Comment (avolkov-intel) [inline sklearnex/basic_statistics/tests/test_basic_statistics.py:417]: We need to extend this test set

File: sklearnex/basic_statistics/tests/test_incremental_basic_statistics.py [hunk 4]
```diff
@@ -450,3 +452,30 @@ def test_results_have_underscores(underscore_first):
...
+)
+@pytest.mark.parametrize(
+    "X_xp",
+    [np, array_api_strict],
+)
+@pytest.mark.parametrize(
+    "w_xp",
+    [np, array_api_strict],
+)
+def test_incbs_weights_in_different_namespace(X_xp, w_xp, with_array_api):
```
Comment (avolkov-intel) [inline sklearnex/basic_statistics/tests/test_incremental_basic_statistics.py:469]: Same here, please add more tests to validate torch/dpnp cpu/gpu

File: onedal/basic_statistics/incremental_basic_statistics.py (hunk 2 above)
Comment (david-cortes-intel) [reply onedal/basic_statistics/incremental_basic_statistics.py:134]: It only takes dtype objects from numpy, so if it's a different class (like array-api-strict), it will error out.

**Review by avolkov-intel — APPROVED** [review]

**Comment by ethanglaser** [comment]
This PR caused github actions CI to go from green to red - please fix in a follow-up

**Comment by ethanglaser** [comment]
I understand that public CI is messy right now. Private CI is currently more stable - please run it on PRs like this.

**Comment by david-cortes-intel** [comment]
> This PR caused github actions CI to go from green to red - please fix in a follow-up

PR with the fix:
https://github.com/uxlfoundation/scikit-learn-intelex/pull/3161

File: onedal/basic_statistics/incremental_basic_statistics.py (hunk 2 above)
Comment (icfaust) [reply onedal/basic_statistics/incremental_basic_statistics.py:134]: @david-cortes-intel this is why we convert to a onedal table first, which will guarantee a numpy dtype.
